### ISSUE TITLE
Compilation error when building Charm++ with "--disable-error-checking"

### ISSUE THREAD

**Issue by bilgeacun** [UIUC-PPL/charm#131]
*Original issue: https://charm.cs.illinois.edu/redmine/issues/131*

---
When building Charm++ with the following commands, Charm++ is NOT built.

* ./build LIBS net-linux-x86_64 --suffix=test-trace --disable-error-checking -j32 -optimize
* ./build LIBS net-linux-x86_64 smp --suffix=test-trace --disable-error-checking -j32 -optimize

build options give this compilation error:

```
debug-charm.C: In function ‘void CpdBeforeEp(int, void*, void*)’:
debug-charm.C:66: error: expected primary-expression before ‘;’ token
../bin/charmc  -optimize  -c -I. NborBaseLB.C
Fatal Error by charmc in directory /home/acun2/charmm/net-linux-x86_64-test-trace/tmp
   Command g++ -m64 -DCMK_GFORTRAN -I../bin/../include -D__CHARMC__=1 -I. -O3 -fno-stack-protector -c debug-charm.C -o debug-charm.o returned error code 1
charmc exiting...
../bin/charmc  -optimize  -c -I. ckgraphTemp.C
gmake: *** [debug-charm.o] Error 1
gmake: *** Waiting for unfinished jobs....
-------------------------------------------------
Charm++ NOT BUILT. Either cd into net-linux-x86_64-test-trace/tmp and try
```

* ./build LIBS net-linux-x86_64 gcc --suffix=test-trace --disable-error-checking -j32 -optimize
* ./build LIBS net-linux-x86_64 gcc smp --suffix=test-trace --disable-error-checking -j32 -optimize

options give the following:
```
debug-charm.C: In function ‘void CpdBeforeEp(int, void*, void*)’:
debug-charm.C:66: error: expected primary-expression before ‘;’ token
../bin/charmc  -optimize  -c -I. ckobjid.C
../bin/charmc  -optimize  -c -I. LBDBManager.C
Fatal Error by charmc in directory /home/acun2/charmm/net-linux-x86_64-gcc-test-trace/tmp
   Command g++ -fPIC -Wno-deprecated -I../bin/../include -D__CHARMC__=1 -I. -O3 -fno-stack-protector -c debug-charm.C -o debug-charm.o returned error code 1
charmc exiting...
../bin/charmc  -optimize  -c -I. LBComm.C
../bin/charmc  -optimize  -c -I. LBObj.C
../bin/charmc  -optimize  -c -I. LBMachineUtil.C
../bin/charmc  -optimize  -c -I. CentralPredictor.C
gmake: *** [debug-charm.o] Error 1
gmake: *** Waiting for unfinished jobs....
-------------------------------------------------
Charm++ NOT BUILT. Either cd into net-linux-x86_64-gcc-test-trace/tmp and try
```

* ./build LIBS net-linux-x86_64 icc --suffix=test-trace --disable-error-checking -j32 -optimize

```
compilation aborted for debug-charm.C (code 2)
Fatal Error by charmc in directory /home/acun2/charmm/net-linux-x86_64-icc-test-trace/tmp
   Command icpc -fpic -I../bin/../include -D__CHARMC__=1 -I. -O3 -fno-stack-protector -c debug-charm.C -o debug-charm.o returned error code 2
charmc exiting...
gmake: *** [debug-charm.o] Error 1
gmake: *** Waiting for unfinished jobs....
ckreduction.C(833): warning #557: pointless comparison of unsigned integer with a negative constant
          if (m->userFlag!=-1)
                         ^

ckreduction.C(1016): warning #557: pointless comparison of unsigned integer with a negative constant
                  if (m->userFlag!=-1)
                                 ^

ckreduction.C(1304): warning #68: integer conversion resulted in a change of sign
    ret->userFlag=-1;
                  ^

ckreduction.C(2245): warning #557: pointless comparison of unsigned integer with a negative constant
          if (m->userFlag!=-1)
                         ^

ckobjid.C(46): warning #1011: missing return statement at end of non-void function "CkObjID::getObject"
  }
  ^

ckobjid.C(68): warning #1011: missing return statement at end of non-void function "CkObjID::guessPE"
  };
  ^

-------------------------------------------------
Charm++ NOT BUILT. Either cd into net-linux-x86_64-icc-test-trace/tmp and try
```

* ./build LIBS net-linux-x86_64 icc smp --suffix=test-trace --disable-error-checking -j32 -optimize

```
../bin/charmc  -optimize  -c -I. ckmessagelogging.C
ckreduction.C(833): warning #557: pointless comparison of unsigned integer with a negative constant
          if (m->userFlag!=-1)
                         ^

ckreduction.C(1016): warning #557: pointless comparison of unsigned integer with a negative constant
                  if (m->userFlag!=-1)
                                 ^

ckreduction.C(1304): warning #68: integer conversion resulted in a change of sign
    ret->userFlag=-1;
                  ^

ckreduction.C(2245): warning #557: pointless comparison of unsigned integer with a negative constant
          if (m->userFlag!=-1)
                         ^

debug-charm.C(66): error: expected an expression
      entry.previousChareID = setMemoryChareIDFromPtr(obj);
                                                          ^

debug-charm.C(1018): warning #2193: null argument provided for parameter marked with attribute "nonnull"
   if (-1 == execvp(gdbScript, NULL))
                               ^

ck.C(2200): warning #654: overloaded virtual function "CkMessageWatcher::process" is only partially overridden in class "CkMessageDetailRecorder"
  class CkMessageDetailRecorder : public CkMessageWatcher {
        ^

ck.C(2431): warning #654: overloaded virtual function "CkMessageWatcher::process" is only partially overridden in class "CkMessageDetailReplay"
  class CkMessageDetailReplay : public CkMessageWatcher {
        ^

ck.C(2622): warning #186: pointless comparison of unsigned integer with zero
      if (e->getSrcPe()<0 || e->getSrcPe()>=CkNumPes()+CkNumNodes()) return 0;
                       ^

compilation aborted for debug-charm.C (code 2)
Fatal Error by charmc in directory /home/acun2/charmm/net-linux-x86_64-smp-icc-test-trace/tmp
   Command icpc -fpic -D_REENTRANT -I../bin/../include -D__CHARMC__=1 -I. -O3 -fno-stack-protector -c debug-charm.C -o debug-charm.o returned error code 2
charmc exiting...
gmake: *** [debug-charm.o] Error 1
gmake: *** Waiting for unfinished jobs....
ckarrayreductionmgr.C(124): warning #557: pointless comparison of unsigned integer with a negative constant
  			if (m->userFlag!=-1)
  			               ^

-------------------------------------------------
Charm++ NOT BUILT. Either cd into net-linux-x86_64-smp-icc-test-trace/tmp and try

```

**Comment by nikhil-jain** [comment]
*Original date: 2013-03-26 02:49:48*

---
This tells us that passing -optimize flag to build does not enable CMK_OPTIMIZE flag. Also, debug-charm.C is using CMK_OPTIMIZE flag for enabling debugging code; using CMK_ERROR_CHECKING may be more accurate. This should affect all builds (break all of them using this flag combination)!

**Comment by PhilMiller** [comment]
*Original date: 2013-03-27 21:47:00*

---
Happens on net-linux-x86_64, too. It was a silly problem of mismatched tests against `CMK_CHARMDEBUG/CMK_ERROR_CHECKING/CMK_OPTIMIZE`. Will fix shortly.

**Comment by PhilMiller** [comment]
*Original date: 2013-03-27 22:10:42*

---
Fixed in commit 96a08a1ba8f5eea783639ee89e52f881a0ea3ebb. Please confirm that compilation now succeeds under the affected options and close.

**Comment by PhilMiller** [comment]
*Original date: 2013-03-27 22:16:53*

---
This is pretty easy to backport, and fixes a brown-paper-bag bug.
